### ISSUE TITLE
Does not focus application if use cmd+tab across screens

### ISSUE THREAD

**Issue by maksimr** [koekeishiya/kwm#224]
UC:

1. Open Google Chrome
2. Open iTerm
3. Focus iTerm
4. Press <cmd>+tab
5. Select Google Chrome

AR:
Google Chrome is selected but it's not focused

ER:
Focus Google Chrome

KWM version -  https://github.com/koekeishiya/kwm/commit/47377480a3bd7ebce385298580ddedf6d04e6c90

NOTE: It's also does not work if select application on another space

http://screencast.com/t/GrQmMLL65V
http://screencast.com/t/dEor0RxILyAG

Thanks!

**Comment by chrishoage** [comment]
cmd + tab works just fine for me. Are you using monocle mode? 

(I don't use Spaces so I can't speak to that issue)

**Comment by koekeishiya** [comment]
cmd + tab worked just fine for me, during testing. Probably an issue with monocle mode.

**Comment by chrishoage** [comment]
I just now noticed that cmd + ` has regressed as far as kwm knowing it gets focus. 

OS X gives the window focus with cmd + ` but kwm doesn't know and still acts as if the previous window has focus.

**Comment by koekeishiya** [comment]
The issue with focusing an application a different space, is 99% related to the fact that I make Kwm refocus the application that had focus on that space.

I suspect what is happening is this:
1. User activates application on different space.
2. OSX triggers space transition to give focus.
3. Kwm notices end of space transition and focuses the application
    that was previously focused on that space.

> I just now noticed that cmd + ` has regressed as far as kwm knowing it gets focus.

> OS X gives the window focus with cmd + ` but kwm doesn't know and still acts as if the previous window has focus.

I am aware of this, but have no idea how to fix this. In theory it simply requires to observe the "AXFocusedWindowChangedNotification", however this triggered twice (for two separate windows) the last time I tried, and caused a lot of issues.

To be honest, these are probably at the bottom of my priority list of things to fix as well, considering I don't use cmd+tab or cmd+`

**Comment by maksimr** [comment]
> cmd + tab works just fine for me. Are you using monocle mode?

Yes

**Comment by maksimr** [comment]
I have checked bsp mode and it work correct for one space but if I try to select application from another space(using cmd+tab) it does not work

**Comment by maksimr** [comment]
> To be honest, these are probably at the bottom of my priority list of things to fix as well, considering I don't use cmd+tab or cmd+`

What are you using for fast switching between application? For example if want open Adium which does not have window by default and is float application.

**Comment by koekeishiya** [comment]
> What are you using for fast switching between application? For example if want open Adium which does not have window by default and is float application.

I very rarely have many applications floating, so I keep them all on a separate space that I have in float mode.

I usually have spaces dedicated for a single purpose. Space 1 would be development. Space 2 webbrowsing / procrastination. Space 3 for chat applications and so on.

I fast switch by using hotkeys to move to specific spaces (ctrl+[num]), or to focus a specific screen (ctrl+alt+[num]).

**Comment by maksimr** [comment]
@koekeishiya thanks, I will try your workflow

**Comment by koekeishiya** [comment]
> cmd + tab works just fine for me. Are you using monocle mode?
(I don't use Spaces so I can't speak to that issue)

So it seems like this is also broken for BSP spaces in the latest commit.

**Comment by koekeishiya** [comment]
Is cmd+tab really important for people? The code that is supposed to fix cmd+tab support, is the reason for #226 which is really annoying. I currently do not see how to have both of them working as expected.

**Comment by chrishoage** [comment]
cmd + tab and friends aren't important to me, but the fact that what is
currently focused and want kwm thinks is focused can get out of sync is a
little bit of an issue for me. In a perfect world kwm would always know
what the focused window of OS X is at all times.

**Comment by koekeishiya** [comment]
> cmd + tab and friends aren't important to me, but the fact that what is currently focused and want kwm thinks is focused can get out of sync is a little bit of an issue for me. In a perfect world kwm would always know what the focused window of OS X is at all times.

I could in theory, make it so they are always synced, however, I suspect this would degrade performance quite a bit. It would also make Kwm update the focus before a window is tiled, which is the reason for #226 in the first place. There are probably other things that could break by doing this.

**Comment by koekeishiya** [comment]
>> In a perfect world kwm would always know what the focused window of OS X is at all times.

> I could in theory, make it so they are always synced, however, I suspect this would degrade performance quite a bit. It would also make Kwm update the focus before a window is tiled, which is the reason for #226 in the first place. There are probably other things that could break by doing this.

I have pushed changes that should make it such that Kwm always has the same focus as OSX, here: https://github.com/koekeishiya/kwm/tree/SyncOSXFocus

This does not degrade performance at all, as far as I can tell, and I find it to be a lot more usable than the current master branch. The problem however is #226. By doing this focus sync, Kwm can no longer use the insert windows at the focused container, as it has already focused the new window before it can try to tile it.

**Comment by chrishoage** [comment]
> This does not degrade performance at all, as far as I can tell, and I find it to be a lot more usable than the current master branch

I will run this branch for the next 24 hours and report back if there are any perf issues. Initially there don't seem to be any, but i've only been suing it for a few minutes. 

The #266 bug is kind of annoying, is there any way to track the previous focused window and use that one, with the knowledge that the focused window is a new window?

**Comment by koekeishiya** [comment]
> The #266 bug is kind of annoying, is there any way to track the previous focused window and use that one, with the knowledge that the focused window is a new window? 

This would be the ideal solution. I need to somehow be able to know when a window is tiled, whether that window was given focus by Kwm, or by OSX.

**Comment by chrishoage** [comment]
>  I need to somehow be able to know when a window is tiled, whether that window was given focus by Kwm, or by OSX.

I'm not saying this is the only way (there is probably a more elegant solution), but couldn't kwm just track, that state in a hash? Kwm has knowledge of the windowID, therefore it could keep a hash table of windowIDs it knows about, has tiled, has focused, etc.

(this is with out understanding of how kwm currently maps that state)

**Comment by koekeishiya** [comment]
I believe I may have solved the issue, need to test it some more.

**Comment by koekeishiya** [comment]
#226 should now be fixed in the latest commit.

**Comment by chrishoage** [comment]
It's half fixed. The behavior is better than before, but kwm doesn't update itself until the mouse is moved / clicked. 

`kwmc read tag` shows as `[0]` for my monocle space until a click or mouse move happens.

**Comment by koekeishiya** [comment]
> It's half fixed. The behavior is better than before, but kwm doesn't update itself until the mouse is moved / clicked.

> kwmc read tag shows as [0] for my monocle space until a click or mouse move happens.

Awkward, will look into it when I have time.
Do the other changes break anything, or does it seem like it could be merged into master?

**Comment by chrishoage** [comment]
I've used it for a little under 24 hours, and it seems to be fine so far. I haven't noticed it break anything so far.

**Comment by chrishoage** [comment]
Odd behavior: On screen 2, clicking a link that opens in screen 1 makes `kwmc screen -f prev` not work. It's as if kwm "forgets" about the screen until I use the mouse to give focus to screen 2 manually.

**Comment by dominiklohmann** [comment]
> Odd behavior: On screen 2, clicking a link that opens in screen 1 makes kwmc screen -f prev not work. It's as if kwm "forgets" about the screen until I use the mouse to give focus to screen 2 manually.

Ohhh that is what causes that. I've been wondering all day why it stops working every so often.

**Comment by chrishoage** [comment]
@dominiklohmann are you on `SyncOSXFocus` or `master`? I assumed the issue was with `SyncOSXFocus` but if you're on `master` then the issue is not caused from the changes on that branch.

**Comment by dominiklohmann** [comment]
I switched to that branch yesterday late night - and kinda forgot about it, wondering all day what causes that behaviour -.-'

**Comment by chrishoage** [comment]
:+1:  cool, good info, seems to be an issue with `SyncOSXFocus` @koekeishiya

**Comment by koekeishiya** [comment]
> Odd behavior: On screen 2, clicking a link that opens in screen 1 makes kwmc screen -f prev not work. It's as if kwm "forgets" about the screen until I use the mouse to give focus to screen 2 manually. 

link as in URL?

**Comment by chrishoage** [comment]
Correct, this behavior might exists for anything else that redirects focus back to the other display. The only thing I have seen cause this so far is clicking a URL that opens in a browser on the other display.

**Comment by chrishoage** [comment]
More odd behavior, I have not tried master, but I can't trigger the prefix key in ONLY safari. 

I have `ctrl + k` bound for my prefix key. The border switches to the prefix color in any app aside from safari. Will investigate this behavior more in a few hours but curious if anyone else is seeing that behavior.

**Comment by koekeishiya** [comment]
The focus changes done on the `Sync` branch has a lot of issues when it comes to multiple monitors, that I do not know how to handle. 

It does work better than `master` in a single monitor setup though.

**Comment by koekeishiya** [comment]
> Correct, this behavior might exists for anything else that redirects focus back to the other display. The only thing I have seen cause this so far is clicking a URL that opens in a browser on the other display. 

This actually happens every single time screen focus changes because of OSX.
`kwmc screen -f id` breaks as well, which is really weird.

**Comment by chrishoage** [comment]
Ok, I'm not taking crazy pills, and its not an issue with safari. Keypress sometimes lag across all applications. They are delivered to the application a second or two after they were issued. 

I have had: `cmd + t` in chrome not respond, so I pressed in 3 times. 2 seconds later 3 tabs opened. Similar issue with safari and the prefix issue.

**Comment by koekeishiya** [comment]
> Ok, I'm not taking crazy pills, and its not an issue with safari. Keypress sometimes lag across all applications. They are delivered to the application a second or two after they were issued.

> I have had: cmd + t in chrome not respond, so I pressed in 3 times. 2 seconds later 3 tabs opened. Similar issue with safari and the prefix issue.

This is not on the master branch, right?

**Comment by chrishoage** [comment]
Correct, I have been on `SyncOSXFocus` for the last 24 hours or so.

**Comment by koekeishiya** [comment]
I'm unable to reproduce the key-delay issues, though I still assume that it is caused by event-tap delays.

**Comment by chrishoage** [comment]
> I'm unable to reproduce the key-delay issues, though I still assume that it is caused by event-tap delays.
Probably related to #92

Yeah once I figured out what was causing it, it is totally #92

**Comment by koekeishiya** [comment]
Try the latest commit on `SyncOSXFocus` and let me know if key-delay issues are less severe, or non-existant.

I've made it so that the Event Tap simply checks if a key is a hotkey, and then adds it to a queue of hotkeys to be processed, which happens on its own thread. This will prevent Kwm from causing delay in the OSX Event Dispatcher, and should in theory fix the delay issue.

Previously, when holding down a hotkey, say, `kwmc window -f west` with `cycle-focus screen`, you would see the lag clearly happening. This is no longer the cause after the latest commit.

I have not been able to reproduce the issues from #92 on my machine before this commit either, so I need someone else to verify if these changes in fact did help.

**Comment by chrishoage** [comment]
> Try the latest commit on SyncOSXFocus and let me know if key-delay issues are less severe, or non-existant.

HEAD on `SyncOSXFocus` makes kwm not work as far as I can tell. `kwm` starts (as evidenced by `kwm read tag`) but no tiling happens, no key commands work, and the border does not function. 

> This will prevent Kwm from causing delay in the OSX Event Dispatcher, and should in theory fix the delay issue.

This sounds like  a pretty good approach

> I have not been able to reproduce the issues from #92 on my machine before this commit either, so I need someone else to verify if these changes in fact did help.

Thats very odd, this happens to me at minimum once a day. I wonder if I have some software that conflicts with kwm in an odd way. I do use BetterTouchTool, which might work on event taps, do you have any knowledge if two programs that subscribe to event taps they would conflict with each other?

Once the branch works I can try it out for 24 hours and report back if I have any more lag issues, as it is a daily occurrence for me.

**Comment by koekeishiya** [comment]
> HEAD on SyncOSXFocus makes kwm not work as far as I can tell. kwm starts (as evidenced by kwm read tag) but no tiling happens, no key commands work, and the border does not function. 

Really? Everything works as normal here

**Comment by chrishoage** [comment]
I'll try building in dev mode and see what happens

**Comment by chrishoage** [comment]
I get no log output in kwm at all when in dev mode? Did dev mode change from `make`?

**Comment by chrishoage** [comment]
Here is what I did 
```
chrishoage at Constance in ~/Projects/kwm on SyncOSXFocus
❯ launchctl unload ~/Library/LaunchAgents/com.koekeishiya.kwm.plist

chrishoage at Constance in ~/Projects/kwm on SyncOSXFocus
❯ make
g++ -c kwm/kwm.cpp -DDEBUG_BUILD -g -O3 -Wall -o obj/kwm/kwm.o
g++ -c kwm/window.cpp -DDEBUG_BUILD -g -O3 -Wall -o obj/kwm/window.o
kwm/window.cpp:1426:5: warning: 'GetProcessForPID' is deprecated: first deprecated in OS X
      10.9 [-Wdeprecated-declarations]
    GetProcessForPID(KWMFocus.Window->PID, &NewPSN);
    ^
/System/Library/Frameworks/ApplicationServices.framework/Frameworks/HIServices.framework/Headers/Processes.h:930:1: note:
      'GetProcessForPID' has been explicitly marked deprecated here
GetProcessForPID(
^
kwm/window.cpp:1466:5: warning: 'GetProcessForPID' is deprecated: first deprecated in OS X
      10.9 [-Wdeprecated-declarations]
    GetProcessForPID(KWMFocus.Window->PID, &NewPSN);
    ^
/System/Library/Frameworks/ApplicationServices.framework/Frameworks/HIServices.framework/Headers/Processes.h:930:1: note:
      'GetProcessForPID' has been explicitly marked deprecated here
GetProcessForPID(
^
kwm/window.cpp:1474:9: warning: 'SetFrontProcessWithOptions' is deprecated: first deprecated
      in OS X 10.9 [-Wdeprecated-declarations]
        SetFrontProcessWithOptions(&KWMFocus.PSN, kSetFrontProcessFrontWindowOnly);
        ^
/System/Library/Frameworks/ApplicationServices.framework/Frameworks/HIServices.framework/Headers/Processes.h:658:1: note:
      'SetFrontProcessWithOptions' has been explicitly marked deprecated here
SetFrontProcessWithOptions(
^
3 warnings generated.
g++ -c kwm/keys.cpp -DDEBUG_BUILD -g -O3 -Wall -o obj/kwm/keys.o
g++ -c kwm/notifications.cpp -DDEBUG_BUILD -g -O3 -Wall -o obj/kwm/notifications.o
g++ -c kwm/workspace.mm -DDEBUG_BUILD -g -O3 -Wall -o obj/kwm/workspace.o
g++ obj/kwm/kwm.o obj/kwm/tree.o obj/kwm/window.o obj/kwm/display.o obj/kwm/daemon.o obj/kwm/interpreter.o obj/kwm/keys.o obj/kwm/space.o obj/kwm/border.o obj/kwm/notifications.o obj/kwm/workspace.o obj/kwm/serializer.o -DDEBUG_BUILD -g -O3 -Wall -lpthread -framework ApplicationServices -framework Carbon -framework Cocoa -o bin/kwm
mkdir -p /Users/chrishoage/.kwm
if test ! -e /Users/chrishoage/.kwm/kwmrc; then cp -n examples/kwmrc /Users/chrishoage/.kwm/kwmrc; fi

chrishoage at Constance in ~/Projects/kwm on SyncOSXFocus
❯ codesign -f -s "Christopher Hoage" ./bin/kwm

chrishoage at Constance in ~/Projects/kwm on SyncOSXFocus
❯ ./bin/kwm
```

Note no output from `kwm`

**Comment by koekeishiya** [comment]
The makefile has been changed to build .obj files, this way it will only re-build files that change.
Doing `make clean` followed by `make` should work just fine.

**Comment by chrishoage** [comment]
After a make clean 
```
chrishoage at Constance in ~/Projects/kwm on SyncOSXFocus
❯ ./bin/kwm
Local Daemon is now running..
```

thats it, nothing works after I start it.

**Comment by koekeishiya** [comment]
Well that is weird, works fine for me every single time i rebuild. Doesn't matter if i do a dev or install build.

**Comment by chrishoage** [comment]
9de488abc8b98c7d92e4b0867fae6327757d74e4 is where it stopped working for me. The commit before that works fine.

**Comment by chrishoage** [comment]
Going for broke and restarting my machine. One moment.

**Comment by chrishoage** [comment]
Crashed on boot. Still can't get it to work in dev mode at all. https://gist.github.com/chrishoage/f29e8f8c56bc997158da

**Comment by koekeishiya** [comment]
Right, found a case that I forgot to fix regarding adding hotkeys. It should work now.

**Comment by chrishoage** [comment]
:+1: it is indeed working now. I'll run this for 24 hours and report back if I have any laggy keypresses still.

**Comment by koekeishiya** [comment]
> :+1: it is indeed working now. I'll run this for 24 hours and report back if I have any laggy keypresses still. 

Cheers, hope the problem is less severe now.

**Comment by koekeishiya** [comment]
> Correct, this behavior might exists for anything else that redirects focus back to the other display. The only thing I have seen cause this so far is clicking a URL that opens in a browser on the other display. 

This should be fixed in the latest commit.

On topic:
The issue in question should also be fixed, please do run the latest commit for a day or so to confirm that it can be merged into master without causing any glaring issues.

I've come to realize that I seem to be using a very small part of what is actually supported, so it is quite hard for me to verify whether things work properly or not.

**Comment by chrishoage** [comment]
> This should be fixed in the latest commit.

It does appear to be fixed.

> The issue in question should also be fixed, please do run the latest commit for a day or so to confirm that it can be merged into master without causing any glaring issues.

The focus-border does indeed update, but at least with a space in monocle mode issuing a focus command (for example `kwmc window -f east`) effects the previously focused window on the previous display. It also breaks `kwmc screen -f next | prev` in the same way the link-opening issue.

**Comment by koekeishiya** [comment]
> The focus-border does indeed update, but at least with a space in monocle mode issuing a focus command (for example kwmc window -f east) effects the previously focused window on the previous display. It also breaks kwmc screen -f next | prev in the same way the link-opening issue. 

I can reproduce this if I do a "slow" cmd+tab (letting the interface show up). A "fast" cmd+tab does not have this problem for some reason.

**Comment by chrishoage** [comment]
> I can reproduce this if I do a "slow" cmd+tab (letting the interface show up). A "fast" cmd+tab does not have this problem for some reason.

Holy crap you're right. That is _so_ strange.

**Comment by koekeishiya** [comment]
> Holy crap you're right. That is so strange. 

I have been able to fix this in the latest commit. The same change is a fix for #184.

**Comment by koekeishiya** [comment]
I believe this is stable enough now. Gonna merge it into master.
